### PR TITLE
nikshay migration tweaks and performance updates

### PR DESCRIPTION
**Pull request by czue**
@proteusvacuum @nickpell will just push to this branch as i make progress and make sure it's always mergeable.

the big change is to chunk everything up in batches which makes it way way faster. f48d2bf

if this still needs additional optimization i can deeper dive into it but wanted to let you guys take it for a spin for now and see if this is good enough.

review by commit. :blowfish:

### CODE REVIEW

File: custom/enikshay/nikshay_datamigration/factory.py [hunk 1]
```diff
@@ -30,25 +35,72 @@ def __init__(self, domain, patient_detail, nikshay_codes_to_location):
...
+        Get the existing occurrence case for this nikshay ID, or None if no occurrence case exists
+        """
+        if self.existing_person_case:
+            try:
+                return get_open_occurrence_case_from_person(
+                    self.domain, self.existing_person_case.case_id
+                )
+            except ENikshayCaseNotFound:
+                return None
+    @property
```
Comment (dimagimon) [inline custom/enikshay/nikshay_datamigration/factory.py]: E301 expected 1 blank line, found 0

File: custom/enikshay/nikshay_datamigration/factory.py [hunk 2]
```diff
@@ -189,23 +226,26 @@ def test(self, followup):
...
-            kwargs['attrs']['create'] = True
+        if self.existing_occurrence_case:
+            matching_test_case = next((
+                extension_case for extension_case in self.case_accessor.get_cases([
+                    index.referenced_id for index in
+                    self.existing_occurrence_case.reverse_indices
+                ])
+                if (
+                    extension_case.type == TEST_CASE_TYPE
+                    and followup.id == int(extension_case.dynamic_case_properties().get('migration_followup_id', -1))
```
Comment (dimagimon) [inline custom/enikshay/nikshay_datamigration/factory.py]: E501 line too long (117 > 115 characters)

**Comment by czue** [comment]
alrighty this is good to go on my side. the big change is to chunk everything up in batches which makes it way way faster. https://github.com/dimagi/commcare-hq/pull/14314/commits/f48d2bf4532bd0c52ab3908c15c0072eec61f086

if this still needs additional optimization i can deeper dive into it but wanted to let you guys take it for a spin for now and see if this is good enough.

File: custom/enikshay/nikshay_datamigration/factory.py [hunk 3]
```diff
@@ -189,23 +226,26 @@ def test(self, followup):
                 relationship=CASE_INDEX_EXTENSION,
                 related_type=occurrence_structure.attrs['case_type'],
             )],
+            # this prevents creating duplicate occurrence data on creation of the test cases
+            'walk_related': False,
```
Comment (czue) [inline custom/enikshay/nikshay_datamigration/factory.py]: this is another notable change. without it we were copying the entire create/update blocks for the person and occurrence in with every test as well. see https://india.commcarehq.org/a/enikshay-test-2/reports/form_data/f5d4eb3a3b714957a8eb3589badd0600/ for an example of what i mean.

**Comment by nickpell** [comment]
this looks great!

running in ```/home/cchq/www/softlayer/releases/2016-12-22_14.23``` on ```db0```.  So far is going much faster!

File: custom/enikshay/nikshay_datamigration/factory.py (hunk 3 above)
Comment (nickpell) [inline custom/enikshay/nikshay_datamigration/factory.py]: oh nice find

**Pull request merged**